add support for running mul invocations in parallel

mirnawong1 opened the pull request:
this pr clarifies that the Cloud CLI now support running multiple invocations in parallel. This is based on @dichenqiandbt 's demo.

Before that cloud CLI only supports run one invocation at one time.

This pr has grown to also address parallel execution, what it means, where it's supported, and modify the current dbt commands table to further explain this. 

Resolves #4952

mirnawong1 committed:
add support for running mul invocations in parallel

this pr clarifies that the Cloud CLI now support running multiple invocations in parallel. This is based on @dichenqiandbt 's demo.

Before that cloud CLI only supports run one invocation at one time.

mirnawong1 commented on website/docs/docs/cloud/configure-cloud-cli.md:
```diff
@@ -95,6 +95,7 @@ To set environment variables in the dbt Cloud CLI for your dbt project:
 ## Use the dbt Cloud CLI
 
 - The dbt Cloud CLI uses the same set of [dbt commands](/reference/dbt-commands) and [MetricFlow commands](/docs/build/metricflow-commands) as dbt Core to execute the commands you provide. For example, use the [`dbt environment`](/reference/commands/dbt-environment) command to view your dbt Cloud configuration details.
+- You can run multiple different invocations or commands in parallel. Note, that you're unable to run the same dbt commands in parallel. (For example, `dbt build` in parallel isn't supported). 
```
```suggestion
- You can run multiple different invocations or commands in parallel. For example, `dbt build` and `dbt parse`.  Note, that you're unable to run the same dbt commands in parallel. For example, running `dbt build` at the same time isn't supported.
```

mirnawong1 committed:
Update website/docs/docs/cloud/configure-cloud-cli.md

dichenqiandbt commented on website/docs/docs/cloud/configure-cloud-cli.md:
```diff
@@ -95,6 +95,7 @@ To set environment variables in the dbt Cloud CLI for your dbt project:
 ## Use the dbt Cloud CLI
 
 - The dbt Cloud CLI uses the same set of [dbt commands](/reference/dbt-commands) and [MetricFlow commands](/docs/build/metricflow-commands) as dbt Core to execute the commands you provide. For example, use the [`dbt environment`](/reference/commands/dbt-environment) command to view your dbt Cloud configuration details.
+- You can run multiple different invocations or commands in parallel. For example, `dbt build` and `dbt parse`.  Note, that you're unable to run the same dbt commands in parallel. For example, running `dbt build` at the same time isn't supported.
```
I hope to clarify a little to make sure the words here is easy for customer to understand.

Today we support run 2 invocations but it might be more in future, depends on user feedback.

dichenqiandbt replied:
It's not 100% accurate, on backend we categorize dbt commands into two types, data warehouse write and data warehouse non-write. 

Data warehouse write command always has 1 parallelism. E.g. build. They may cause data warehouse confliction, e.g. overwrite the same table.
Data warehouse non-write command can have x(today it's 1, but it might be more, let's say 2) parallelism. E.g. parse. They are safe to run in parallel.

I'm not sure how to phrase this as it's too complicated for customer to understand, maybe say you are not able to run data warehouse conflicted commands?

dichenqiandbt commented on website/docs/docs/cloud/configure-cloud-cli.md:
```diff
@@ -95,6 +95,7 @@ To set environment variables in the dbt Cloud CLI for your dbt project:
 ## Use the dbt Cloud CLI
 
 - The dbt Cloud CLI uses the same set of [dbt commands](/reference/dbt-commands) and [MetricFlow commands](/docs/build/metricflow-commands) as dbt Core to execute the commands you provide. For example, use the [`dbt environment`](/reference/commands/dbt-environment) command to view your dbt Cloud configuration details.
+- You can run multiple different invocations or commands in parallel. For example, `dbt build` and `dbt parse`.  Note, that you're unable to run the same dbt commands in parallel. For example, running `dbt build` at the same time isn't supported.
```
Curious is it intended to have two whitespace before "note"?

mirnawong1 replied:
nice catch, thank you so much!

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

Commit:
fold feedback

Commit:
Merge branch 'mirnawong1-patch-22' of https://github.com/dbt-labs/docs.getdbt.com into mirnawong1-patch-22

mirnawong1 commented on website/docs/docs/cloud/configure-cloud-cli.md:
```diff
@@ -104,3 +104,13 @@ As a tip, most command-line tools have a `--help` flag to show available command
+
+- **Data platform write commands** &mdash; Commands such as `dbt build` and `dbt run` that perform write operations to your data platform. These commands are limited to one invocation at any given time. This is to prevent any potential conflicts, such as overwriting the same table in your data platform, at the same time. For example, you can't run `dbt build` and `dbt run` at the same time.
+
+- **Data platform read commands** &mdash; Commands such as `dbt parse` and `dbt source snapshot-freshness` that don't write to your platform. These commands aren't limited to one invocation at any given time and you can run multiple invocations in parallel. For example, you can run `dbt parse` and `dbt source snapshot-freshness` at the same time.
```
@dichenqiandbt do we want to say 'read' or is 'non-write' the correct way to phrase this? I've opted for 'read'.

dichenqiandbt replied:
A little bit tricky here, as some commands won't interactive with data warehouse, e.g. deps, docs. Technically write, read only and data warehouse irrelevant, three types of commands.

But I'm neutral on this, if read is better for customer to understand, that's totally fine.

mirnawong1 commented on website/docs/docs/cloud/configure-cloud-cli.md:
```diff
@@ -104,3 +104,13 @@ As a tip, most command-line tools have a `--help` flag to show available command
 - `dbt run --help`: Lists the flags available for the `run` command
 :::
+
+### Run dbt commands
```
@dichenqiandbt is the categorization (write and non write) applicable to all of dbt (core and cloud) or is this behavior only for the cloud cli?

dichenqiandbt replied:
For all, it's a very general categorization.

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

dichenqiandbt commented on website/docs/docs/cloud/configure-cloud-cli.md:
```diff
@@ -104,3 +104,13 @@ As a tip, most command-line tools have a `--help` flag to show available command
+
+- **Data platform write commands** &mdash; Commands such as `dbt build` and `dbt run` that perform write operations to your data platform. These commands are limited to one invocation at any given time. This is to prevent any potential conflicts, such as overwriting the same table in your data platform, at the same time. For example, you can't run `dbt build` and `dbt run` at the same time.
+
+- **Data platform read commands** &mdash; Commands such as `dbt parse` and `dbt source snapshot-freshness` that don't write to your platform. These commands aren't limited to one invocation at any given time and you can run multiple invocations in parallel. For example, you can run `dbt parse` and `dbt source snapshot-freshness` at the same time.
```
Cloud CLI do have a parallelism limit on non-read commands, can you rephrase by adding a "ideally aren't limited to...". 

Today our limit is 1 for non-write invocations but it might be increased soon, so I don't want to say a concrete limit number here, otherwise it would be updated couple of times.

LMK if my expression is clear.

Commit:
updates

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

Commit:
updates

Commit:
update

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

dbeatty10 commented on website/docs/reference/dbt-commands.md:
```diff
@@ -99,3 +98,47 @@ Use the following dbt commands in [dbt Core](/docs/core/installation-overview) a
+
+Note that some dbt commands aren't listed in the following table since they're not relevant to the parallelization of dbt commands.
+
+| Command          | Type       | Parallel execution <br /> with other commands |
+| :-----------------: | :----------:|:------------------:|
+| `build`          | Write  |     ❌             |
+| `clone`          | Write  |     ❌              |
+| `retry`          | Write  |     ❌             |
+| `run`            | Write  |     ❌             |
+| `run-operation`  | Write  |     ❌             |
+| `seed`           | Write  |     ❌             |
+| `snapshot`       | Write  |     ❌             |
+| `clean`          | Read |      ✅           |
+| `compile`        | Read |      ✅           |
+| `debug`          | Read |      ✅           |
+| `deps`           | Read |      ✅           |
+| `docs`           | Read |      ✅           |
+| `init`           | Read |      ✅           |
+| `list`           | Read |      ✅           |
+| `parse`          | Read |      ✅           |
+| `show`           | Read |      ✅           |
+| `source`         | Read |     ✅           |
+| `test`           | Read |     ✅           |
```
What do you think about adding the "parallel execution" column to the 
[available commands](https://docs-getdbt-owu0eidj8-dbt-labs.vercel.app/reference/dbt-commands#available-commands) table instead?

<img width="400" alt="image" src="https://github.com/dbt-labs/docs.getdbt.com/assets/44704949/67774541-17fd-4524-b8ac-9bd66e2b1071">

mirnawong1 replied:
i did tihnk about this too @dbeatty10 but I stopped short of it bc i thought it was looking too 'busy' - wdyt? 
![Screenshot 2024-02-20 at 14 24 04](https://github.com/dbt-labs/docs.getdbt.com/assets/89008547/63dc3e0f-f7a0-4d8f-b4a1-dc3e987b61fa)

jtcohen6 replied:
I like this idea!

mirnawong1 replied:
ok ya'll going to do this!

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

jtcohen6 commented on website/docs/docs/cloud/configure-cloud-cli.md:
```diff
@@ -94,9 +94,11 @@ To set environment variables in the dbt Cloud CLI for your dbt project:
-- It also supports [project dependencies](/docs/collaborate/govern/project-dependencies), which allows you to depend on another project using the metadata service in dbt Cloud. 
+The dbt Cloud CLI uses the same set of [dbt commands](/reference/dbt-commands) and [MetricFlow commands](/docs/build/metricflow-commands) as dbt Core to execute the commands you provide. For example, use the [`dbt environment`](/reference/commands/dbt-environment) command to view your dbt Cloud configuration details. With the dbt Cloud CLI, you can:
+
+- Run [multiple invocations in parallel](/reference/dbt-commands#execute-commands) and ensures safe parallelism.
```
Coming from https://github.com/dbt-labs/docs.getdbt.com/issues/4952 - I think we should:
- add a new section to "Programmatic invocations" about how running multiple `dbt-core` invocations simultaneously in the same process is currently NOT safe, officially discouraged, and requires a wrapping process/application to handle sub-processes
- link to that section from here ("ensure safe parallelism, which is not currently guaranteed by dbt-core")
```suggestion
- Run [multiple invocations in parallel](/reference/dbt-commands#execute-commands) and ensure safe parallelism.
```

mirnawong1 replied:
oh great idea! thank you @jtcohen6 will do this!

Commit:
updates

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

Commit:
updates

dbeatty10 commented on website/docs/reference/dbt-commands.md:
```diff
@@ -9,17 +9,43 @@ You can run dbt using the following tools:
 The following sections outline the commands supported by dbt and their relevant flags. For information about selecting models on the command line, consult the docs on [Model selection syntax](/reference/node-selection/syntax).
 
-### Available commands
+## Available commands
+
+| Command | Description | <div style={{width:'200px'}}>Compatible tools and version</div>  | Type | Parallel execution with other commands |
+|---------|-------------|-------------------------------|------|----------------------------------------|
+| [build](/reference/commands/build) | Build and test all selected resources (models, seeds, snapshots, tests) | All tools <br /> All [supported versions](/docs/dbt-versions/core) | Write | ❌ |
+| cancel | Cancels the most recent invocation. | dbt Cloud CLI <br /> Requires [dbt v1.6 or higher](/docs/dbt-versions/core) | N/A | N/A |
+| [clean](/reference/commands/clean) | Deletes artifacts present in the dbt project | All tools <br /> All [supported versions](/docs/dbt-versions/core) | Read | ✅ |
+| [clone](/reference/commands/clone) | Clone selected models from the specified state | All tools <br /> Requires [dbt v1.6 or higher](/docs/dbt-versions/core) | Write | ❌ |
+| [compile](/reference/commands/compile) | Compiles (but does not run) the models in a project | All tools <br /> All [supported versions](/docs/dbt-versions/core) | Read | ✅ |
+| [debug](/reference/commands/debug) | Debugs dbt connections and projects | dbt Cloud IDE, dbt Core <br /> All [supported versions](/docs/dbt-versions/core) | Read | ✅ |
+| [deps](/reference/commands/deps) | Downloads dependencies for a project | All tools <br /> All [supported versions](/docs/dbt-versions/core) | Read | ✅ |
+| [docs](/reference/commands/cmd-docs) | Generates documentation for a project | All tools <br /> All [supported versions](/docs/dbt-versions/core) | Read | ✅ |
+| [environment](/reference/commands/dbt-environment) | Enables you to interact with your dbt Cloud environment. | dbt Cloud CLI <br /> Requires [dbt v1.5 or higher](/docs/dbt-versions/core) | N/A | N/A |
+| help | Displays help information for any command | dbt Core, dbt Cloud CLI <br /> All tools [supported versions](/docs/dbt-versions/core) | N/A | N/A |
+| [init](/reference/commands/init) | Initializes a new dbt project | dbt Core<br /> All [supported versions](/docs/dbt-versions/core) | Read | ✅ |
+| [list](/reference/commands/list) | Lists resources defined in a dbt project | All tools <br /> All [supported versions](/docs/dbt-versions/core) | Read | ✅ |
+| [parse](/reference/commands/parse) | Parses a project and writes detailed timing info | All tools <br /> All [supported versions](/docs/dbt-versions/core) | Read | ✅ |
+| reattach | Reattaches to the most recent invocation to retrieve logs and artifacts. | dbt Cloud CLI <br /> Requires [dbt v1.6 or higher](/docs/dbt-versions/core) | N/A | N/A |
+| [retry](/reference/commands/retry) | Retry the last run `dbt` command from the point of failure | All tools <br /> Requires [dbt v1.6 or higher](/docs/dbt-versions/core) | Write | ❌ |
+| [run](/reference/commands/run) | Runs the models in a project | All tools <br /> All [supported versions](/docs/dbt-versions/core) | Write | ❌ |
+| [run-operation](/reference/commands/run-operation) | Invoke a macro, including running arbitrary maintenance SQL against the database | All tools <br /> All [supported versions](/docs/dbt-versions/core) | Write | ❌ |
+| [seed](/reference/commands/seed) | Loads CSV files into the database | All tools <br /> All [supported versions](/docs/dbt-versions/core) | Write | ❌ |
+| [show](/reference/commands/show) | Preview table rows post-transformation | All tools <br /> All [supported versions](/docs/dbt-versions/core) | Read | ✅ |
+| [snapshot](/reference/commands/snapshot) | Executes "snapshot" jobs defined in a project | All tools <br /> All [supported versions](/docs/dbt-versions/core) | Write | ❌ |
+| [source](/reference/commands/source) | Provides tools for working with source data (including validating that sources are "fresh") | All tools<br /> All [supported versions](/docs/dbt-versions/core) | Read | ✅ |
+| [test](/reference/commands/test)
+
```
What about the following?


The following sections outline the commands supported by dbt and their relevant flags. They are available in all tools and all [supported versions](https://docs.getdbt.com/docs/dbt-versions/core) unless noted otherwise.

|    Command    |                                         Description                                         | Parallel execution with other commands |                Caveats                |
|:-------------:|:-------------------------------------------------------------------------------------------:|:--------------------------------------:|:-------------------------------------:|
| build         | Build and test all selected resources (models, seeds, snapshots, tests)                     | ❌                                      |                                       |
| cancel        | Cancels the most recent invocation.                                                         | N/A                                    | dbt Cloud CLI only dbt v1.6 or higher |
| clean         | Deletes artifacts present in the dbt project                                                | ✅                                      |                                       |
| clone         | Clone selected models from the specified state                                              | ❌                                      | dbt v1.6 or higher                    |
| compile       | Compiles (but does not run) the models in a project                                         | ✅                                      |                                       |
| debug         | Debugs dbt connections and projects                                                         | ✅                                      | dbt Cloud IDE, dbt Core only          |
| deps          | Downloads dependencies for a project                                                        | ✅                                      |                                       |
| docs          | Generates documentation for a project                                                       | ✅                                      |                                       |
| environment   | Enables you to interact with your dbt Cloud environment.                                    | N/A                                    | dbt Cloud CLI only dbt v1.5 or higher |
| help          | Displays help information for any command                                                   | N/A                                    |                                       |
| init          | Initializes a new dbt project                                                               | ✅                                      | dbt Core only                         |
| list          | Lists resources defined in a dbt project                                                    | ✅                                      |                                       |
| parse         | Parses a project and writes detailed timing info                                            | ✅                                      |                                       |
| reattach      | Reattaches to the most recent invocation to retrieve logs and artifacts.                    | N/A                                    | dbt Cloud CLI only dbt v1.6 or higher |
| retry         | Retry the last run dbt command from the point of failure                                    | ❌                                      | dbt v1.6 or higher                    |
| run           | Runs the models in a project                                                                | ❌                                      |                                       |
| run-operation | Invoke a macro, including running arbitrary maintenance SQL against the database            | ❌                                      |                                       |
| seed          | Loads CSV files into the database                                                           | ❌                                      |                                       |
| show          | Preview table rows post-transformation                                                      | ✅                                      |                                       |
| snapshot      | Executes "snapshot" jobs defined in a project                                               | ❌                                      |                                       |
| source        | Provides tools for working with source data (including validating that sources are "fresh") | ✅                                      |                                       |

hash-data commented on website/docs/reference/dbt-commands.md:
```diff
@@ -99,3 +124,50 @@ Use the following dbt commands in [dbt Core](/docs/core/installation-overview) a
+| `seed`           | Write  |     ❌             |
+| `snapshot`       | Write  |     ❌             |
+| `clean`          | Read |      ✅           |
+| `compile`        | Read |      ✅           |
```
but i tested compile multiple times and found it also gives adapter error.

mirnawong1 replied:
Hey @hash-data , can you say more? Would love to see if this is something else or if it's indeed not supported. If it's something more, I would recommend opening an issue in the dbt core repo, however would love more context to help you out, thank you!

hash-data replied:
Hi @mirnawong1  i mentioned the issue here:  https://github.com/dbt-labs/dbt-core/issues/9603
-> Issue i have metioned different commands 
-> But also I tried multiple compile comands which give the same KeyError

jtcohen6 replied:
@hash-data @mirnawong1 This is why it's important we make a distinction between what's supported with:
- dbt-core (CLI or programmatic invocation), which does not support concurrency within the same process (but it's possible to achieve this with additional work)
- Cloud CLI, which does that additional work to manage concurrency (multiple processes) on the user's behalf

(more conversation in https://github.com/dbt-labs/docs.getdbt.com/issues/4952)

Commit:
updates

Commit:
Merge branch 'mirnawong1-patch-22' of https://github.com/dbt-labs/docs.getdbt.com into mirnawong1-patch-22

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

Commit:
folde in feedback

Commit:
Merge branch 'mirnawong1-patch-22' of https://github.com/dbt-labs/docs.getdbt.com into mirnawong1-patch-22

mirnawong1 commented on website/docs/docs/cloud/configure-cloud-cli.md:
```diff
@@ -94,9 +94,11 @@ To set environment variables in the dbt Cloud CLI for your dbt project:
-- It also supports [project dependencies](/docs/collaborate/govern/project-dependencies), which allows you to depend on another project using the metadata service in dbt Cloud. 
+The dbt Cloud CLI uses the same set of [dbt commands](/reference/dbt-commands) and [MetricFlow commands](/docs/build/metricflow-commands) as dbt Core to execute the commands you provide. For example, use the [`dbt environment`](/reference/commands/dbt-environment) command to view your dbt Cloud configuration details. With the dbt Cloud CLI, you can:
+
+- Run [multiple invocations in parallel](/reference/dbt-commands#execute-commands) and ensure [safe parallelism](/reference/dbt-commands#parallel-execution), which is currently not guaranteed by `dbt-core`.
```
do we use `dbt-core` or should we use dbt Core?

jtcohen6 replied:
I prefer `dbt-core` here, since it is a limitation within that specific Python package. I tend to use "dbt Core" more expansively to refer to dbt OSS (including adapters, sometimes also packages).

mirnawong1 replied:
bloody hell you're fast @jtcohen6 🤣! ok thanks for clarifying that - makes complete sense!

Commit:
fix link

Commit:
update to code

dbeatty10 commented on website/docs/reference/dbt-commands.md:
```diff
@@ -5,45 +5,60 @@ title: "dbt Command reference"
+
+| Command type | Description | <div style={{width:'200px'}}>Example</div> |
+|------|-------------|---------|
+| **Write** | These commands perform actions that change data or metadata in your data platform.<br /><br /> Limited to one invocation at any given time (1 parallelism). This is to prevent any potential conflicts, such as overwriting the same table in your data platform, at the same time. | `dbt build`<br />`dbt run` |
```
```suggestion
| **Write** | These commands perform actions that change data or metadata in your data platform.<br /><br /> Limited to one invocation at any given time. This is to prevent any potential conflicts, such as overwriting the same table in your data platform, at the same time. | `dbt build`<br />`dbt run` |
```

dbeatty10 commented on website/docs/reference/dbt-commands.md:
```diff
@@ -5,45 +5,60 @@ title: "dbt Command reference"
+| Command type | Description | <div style={{width:'200px'}}>Example</div> |
+|------|-------------|---------|
+| **Write** | These commands perform actions that change data or metadata in your data platform.<br /><br /> Limited to one invocation at any given time (1 parallelism). This is to prevent any potential conflicts, such as overwriting the same table in your data platform, at the same time. | `dbt build`<br />`dbt run` |
+| **Read** | These commands involve operations that fetch or read data without making any changes to your data platform.<br /><br /> Can have multiple invocations in parallel and aren't limited to one invocation at any given time. This means read commands can run in parallel with other read commands and write commands.| `dbt parse`<br />`dbt compile`|
```
```suggestion
| **Read** | These commands involve operations that fetch or read data without making any changes to your data platform.<br /><br /> Can have multiple invocations in parallel and aren't limited to one invocation at any given time. This means read commands can run in parallel with other read commands and a single write command.| `dbt parse`<br />`dbt compile`|
```

dbeatty10 commented on website/docs/reference/dbt-commands.md:
```diff
@@ -5,45 +5,60 @@ title: "dbt Command reference"
+| [test](/reference/commands/test) | Executes tests defined in a project  |  Read | ✅ | All tools <br /> All [supported versions](/docs/dbt-versions/core) |
+| [--version](/reference/commands/version) | Displays the currently installed version of dbt CLI |  N/A  | N/A | dbt Core, dbt Cloud CLI  <br />  All [supported versions](/docs/dbt-versions/core) | <br />
+
+Note that some have 'N/A" since they're not relevant to the parallelization of dbt commands.
```
```suggestion
Note that some have "N/A" since they're not relevant to the parallelization of dbt commands.
```

mirnawong1 replied:
great catch!

dbeatty10 commented on website/docs/reference/dbt-commands.md:
```diff
@@ -5,45 +5,60 @@ title: "dbt Command reference"
 
 - In your browser with the [dbt Cloud IDE](/docs/cloud/dbt-cloud-ide/develop-in-the-cloud) 
-- On the command line interface using the [dbt Cloud CLI](/docs/cloud/cloud-cli-installation) or open-source [dbt Core](/docs/core/installation-overview), both of which enable you to execute dbt commands. The key distinction is the dbt Cloud CLI is tailored for dbt Cloud's infrastructure and integrates with all its [features](/docs/cloud/about-cloud/dbt-cloud-features).
+- On the command line interface using the [dbt Cloud CLI](/docs/cloud/cloud-cli-installation) or open-source [dbt Core](/docs/core/installation-overview), both of which enable you to execute dbt commands.
```
```suggestion
- On the command line interface using the [dbt Cloud CLI](/docs/cloud/cloud-cli-installation) or open-source [dbt Core](/docs/core/installation-overview).
```

dbeatty10 reviewed (commented):
Handful of suggestions.

mirnawong1 committed:
Update website/docs/reference/dbt-commands.md

Co-authored-by: Doug Beatty <44704949+dbeatty10@users.noreply.github.com>

mirnawong1 committed:
Update website/docs/reference/dbt-commands.md

Co-authored-by: Doug Beatty <44704949+dbeatty10@users.noreply.github.com>

mirnawong1 committed:
Update website/docs/reference/dbt-commands.md

Co-authored-by: Doug Beatty <44704949+dbeatty10@users.noreply.github.com>

mirnawong1 committed:
Update website/docs/reference/dbt-commands.md

Co-authored-by: Doug Beatty <44704949+dbeatty10@users.noreply.github.com>

dbeatty10 reviewed (commented):
Three suggested changes to the table:
1. Omit the "Type" column since it carries the same info as the "Parallel execution" column
1. Omit content in the "Caveat" section if there are no tool or version restrictions. Remove "Requires". Add "only".
1. Replace "N/A" with ✅ if the command can be executed simultaneously (even if it doesn't interact with the database via a read or a write).

This table is exactly the same as the current table. It just . It also 

This table is exactly the same as the current table. It just omits the "Type" column since it carries the same info as the "Parallel execution" column.

It also omits content in the "Caveat" section if there are no tool or version restrictions -- the common assumption is that a command is available in all versions and all tools unless stated otherwise.

There are two versions of the table below:
1. Both suggested changes
2. Omission of "Type" column only

Both are exactly the same as the current table, just less busy.

I didn't do a prototype of suggestion 3, but my _assumption_ is that all the "N/A" could just be replaced with ✅.

### Both suggested changes

| Command | Description | Parallel execution | Caveats |
|---------|-------------------------------| :-----: | ------------------------------|
| [build](/reference/commands/build) | Build and test all selected resources (models, seeds, snapshots, tests) | ❌ |   | 
| cancel | Cancels the most recent invocation. | N/A | dbt Cloud CLI only <br /> [dbt v1.6 or higher](/docs/dbt-versions/core) |
| [clean](/reference/commands/clean) | Deletes artifacts present in the dbt project | ✅ |   |
| [clone](/reference/commands/clone) | Clone selected models from the specified state | ❌ | <br /> [dbt v1.6 or higher](/docs/dbt-versions/core) |
| [compile](/reference/commands/compile) | Compiles (but does not run) the models in a project | ✅ |   |
| [debug](/reference/commands/debug) | Debugs dbt connections and projects | ✅ | dbt Cloud IDE and dbt Core only <br />  |
| [deps](/reference/commands/deps) | Downloads dependencies for a project | ✅ |    |
| [docs](/reference/commands/cmd-docs) | Generates documentation for a project | ✅ |   |
| [environment](/reference/commands/dbt-environment) | Enables you to interact with your dbt Cloud environment. | N/A | dbt Cloud CLI only <br /> [dbt v1.5 or higher](/docs/dbt-versions/core) |
| help | Displays help information for any command | N/A | dbt Core and dbt Cloud CLI  only<br />  |
| [init](/reference/commands/init) | Initializes a new dbt project | ✅ | dbt Core only<br />  |
| [list](/reference/commands/list) | Lists resources defined in a dbt project | ✅ |   |
| [parse](/reference/commands/parse) | Parses a project and writes detailed timing info | ✅ |   |
| reattach | Reattaches to the most recent invocation to retrieve logs and artifacts. | N/A | dbt Cloud CLI only <br /> [dbt v1.6 or higher](/docs/dbt-versions/core) |
| [retry](/reference/commands/retry) | Retry the last run `dbt` command from the point of failure | ❌ |  <br /> [dbt v1.6 or higher](/docs/dbt-versions/core) |
| [run](/reference/commands/run) | Runs the models in a project | ❌ |   |
| [run-operation](/reference/commands/run-operation) | Invoke a macro, including running arbitrary maintenance SQL against the database | ❌ |   |
| [seed](/reference/commands/seed) | Loads CSV files into the database | ❌ |   |
| [show](/reference/commands/show) | Preview table rows post-transformation | ✅ |     |
| [snapshot](/reference/commands/snapshot) | Executes "snapshot" jobs defined in a project | ❌ |   |
| [source](/reference/commands/source) | Provides tools for working with source data (including validating that sources are "fresh") | ✅ | <br />  |
| [test](/reference/commands/test) | Executes tests defined in a project  | ✅ |   |
| [--version](/reference/commands/version) | Displays the currently installed version of dbt CLI | N/A | dbt Core and dbt Cloud CLI only <br />  | <br />


### Omission of "Type" column only

| Command | Description | Parallel execution | Caveats |
|---------|-------------------------------| :-----: | ------------------------------|
| [build](/reference/commands/build) | Build and test all selected resources (models, seeds, snapshots, tests) | ❌ | All tools <br /> All [supported versions](/docs/dbt-versions/core) | 
| cancel | Cancels the most recent invocation. | N/A | dbt Cloud CLI <br /> Requires [dbt v1.6 or higher](/docs/dbt-versions/core) |
| [clean](/reference/commands/clean) | Deletes artifacts present in the dbt project | ✅ | All tools <br /> All [supported versions](/docs/dbt-versions/core) |
| [clone](/reference/commands/clone) | Clone selected models from the specified state | ❌ | All tools <br /> Requires [dbt v1.6 or higher](/docs/dbt-versions/core) |
| [compile](/reference/commands/compile) | Compiles (but does not run) the models in a project | ✅ | All tools <br /> All [supported versions](/docs/dbt-versions/core) |
| [debug](/reference/commands/debug) | Debugs dbt connections and projects | ✅ | dbt Cloud IDE, dbt Core <br /> All [supported versions](/docs/dbt-versions/core) |
| [deps](/reference/commands/deps) | Downloads dependencies for a project | ✅ |  All tools <br /> All [supported versions](/docs/dbt-versions/core) |
| [docs](/reference/commands/cmd-docs) | Generates documentation for a project | ✅ | All tools <br /> All [supported versions](/docs/dbt-versions/core) |
| [environment](/reference/commands/dbt-environment) | Enables you to interact with your dbt Cloud environment. | N/A | dbt Cloud CLI <br /> Requires [dbt v1.5 or higher](/docs/dbt-versions/core) |
| help | Displays help information for any command | N/A | dbt Core, dbt Cloud CLI <br /> All [supported versions](/docs/dbt-versions/core) |
| [init](/reference/commands/init) | Initializes a new dbt project | ✅ | dbt Core<br /> All [supported versions](/docs/dbt-versions/core) |
| [list](/reference/commands/list) | Lists resources defined in a dbt project | ✅ | All tools <br /> All [supported versions](/docs/dbt-versions/core) |
| [parse](/reference/commands/parse) | Parses a project and writes detailed timing info | ✅ | All tools <br /> All [supported versions](/docs/dbt-versions/core) |
| reattach | Reattaches to the most recent invocation to retrieve logs and artifacts. | N/A | dbt Cloud CLI <br /> Requires [dbt v1.6 or higher](/docs/dbt-versions/core) |
| [retry](/reference/commands/retry) | Retry the last run `dbt` command from the point of failure | ❌ | All tools <br /> Requires [dbt v1.6 or higher](/docs/dbt-versions/core) |
| [run](/reference/commands/run) | Runs the models in a project | ❌ | All tools <br /> All [supported versions](/docs/dbt-versions/core) |
| [run-operation](/reference/commands/run-operation) | Invoke a macro, including running arbitrary maintenance SQL against the database | ❌ | All tools <br /> All [supported versions](/docs/dbt-versions/core) |
| [seed](/reference/commands/seed) | Loads CSV files into the database | ❌ | All tools <br /> All [supported versions](/docs/dbt-versions/core) |
| [show](/reference/commands/show) | Preview table rows post-transformation | ✅ |  All tools <br /> All [supported versions](/docs/dbt-versions/core) |
| [snapshot](/reference/commands/snapshot) | Executes "snapshot" jobs defined in a project | ❌ | All tools <br /> All [supported versions](/docs/dbt-versions/core) |
| [source](/reference/commands/source) | Provides tools for working with source data (including validating that sources are "fresh") | ✅ | All tools<br /> All [supported versions](/docs/dbt-versions/core) |
| [test](/reference/commands/test) | Executes tests defined in a project  | ✅ | All tools <br /> All [supported versions](/docs/dbt-versions/core) |
| [--version](/reference/commands/version) | Displays the currently installed version of dbt CLI | N/A | dbt Core, dbt Cloud CLI  <br />  All [supported versions](/docs/dbt-versions/core) | <br />

mirnawong1 commented:
thank you so much for this detailed explanation! I love your suggestions and my instincts go to having all rows under 'caveats' filled in so it's explicit what version and tool is supported ( as opposed to inferred). I'll always opt to be more explicit. I'll change this up and revert back!

dbeatty10 reviewed (approved):
Nice work @mirnawong1 -- I feel much smarter after reading this PR 🧠

dichenqiandbt reviewed (approved):
LGTM, thanks! I like this PR and customer will like it too!

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

Commit:
udpate table

mirnawong1 commented:
updated table to remove the write column but also added a notation to clarify what x and check mean
![Screenshot 2024-02-21 at 18 06 11](https://github.com/dbt-labs/docs.getdbt.com/assets/89008547/74aecd06-fffe-418d-a2d1-3c63dcec8d2d)

runleonarun commented on website/docs/reference/dbt-commands.md:
```diff
@@ -5,45 +5,62 @@ title: "dbt Command reference"
-<VersionBlock firstVersion="1.6">
+In contrast, [`dbt-core` _doesn't_ support](/reference/programmatic-invocations#parallel-execution-not-supported) safe parallel execution for multiple invocations in the same process, and requires users to manage concurrency manually to ensure data integrity and system stability.
+
+dbt commands are categorized into the following types:
```
```suggestion
dbt commands can be `read` or `write` commands:
```

runleonarun commented on website/docs/reference/dbt-commands.md:
```diff
@@ -5,45 +5,62 @@ title: "dbt Command reference"
+
+| Command type | Description | <div style={{width:'200px'}}>Example</div> |
+|------|-------------|---------|
+| **Write** | These commands perform actions that change data or metadata in your data platform.<br /><br /> Limited to one invocation at any given time. This is to prevent any potential conflicts, such as overwriting the same table in your data platform, at the same time. | `dbt build`<br />`dbt run` |
```
```suggestion
| **Write** | These commands perform actions that change data or metadata in your data platform.<br /><br /> Limited to one invocation at any given time, which prevents any potential conflicts, such as overwriting the same table in your data platform at the same time. | `dbt build`<br />`dbt run` |
```

runleonarun commented on website/docs/reference/dbt-commands.md:
```diff
@@ -5,45 +5,62 @@ title: "dbt Command reference"
-| [source](/reference/commands/source) | Provides tools for working with source data (including validating that sources are "fresh")  | All | All [supported versions](/docs/dbt-versions/core) |
-| [test](/reference/commands/test) | Executes tests defined in a project  | All | All [supported versions](/docs/dbt-versions/core) |
-| [--version](/reference/commands/version) | Displays the currently installed version of dbt CLI | dbt Core <br /> dbt Cloud CLI | All [supported versions](/docs/dbt-versions/core) |
+To ensure your dbt workflows are both efficient and safe, you can run different types of dbt commands at the same time (in parallel) &mdash; for example, `dbt build` (write operation) can safely run alongside `dbt parse` (read operation) at the same time. However, you can't run `dbt build` and `dbt run` (both write operations) at the same time.
```
non blocking suggestion. If you move this example before the table, it might feel more cohesive with the other content.

runleonarun reviewed (approved):
Just some small wording suggestions! Ship when you're ready!

mirnawong1 committed:
Update website/docs/reference/dbt-commands.md

Co-authored-by: Leona B. Campbell <3880403+runleonarun@users.noreply.github.com>

mirnawong1 committed:
Update website/docs/reference/dbt-commands.md

Co-authored-by: Leona B. Campbell <3880403+runleonarun@users.noreply.github.com>

mirnawong1 committed:
Update dbt-commands.md

mirnawong1 committed:
Merge branch 'current' into mirnawong1-patch-22

mirnawong1 commented:
thanks for everyone's feedback, i really appreciate it! I'm merging this now!
